docs/design/design_trivy_file_system_scanner.md

```diff
@@ -0,0 +1,197 @@
+# Scan Container Images with Trivy Filesystem Scanner
+
+Authors: [Devendra Turkar], [Daniel Pacak]
+
+## Overview
+
+Starboard currently uses Trivy in [Standalone] or [ClientServer] mode to scan and generate VulnerabilityReports for
+container images by pulling the images from remote registries. Starboard scans a specified K8s workload by running the
+Trivy executable as a K8s Job. This approach implies that Trivy does not have access to images cached by the container
+runtime on cluster nodes. Therefore, to scan images from private registries Starboard reads ImagePullSecrets specified
+on workloads or on service accounts used by the workloads, and passes them down to Trivy executable as `TRIVY_USERNAME`
+and `TRIVY_PASSWORD` environment variables.
+
+Since ImagePullSecrets are not the only way to provide registry credential, the following alternatives are not
+currently supported by Starboard:
+1. Pre-pulled images
+2. [Configuring nodes to authenticate to a private registry]
+3. Vendor-specific or local extension. For example, methods described on [AWS ECR Private registry authentication].
+
+Even though we could resolve some of above-mentioned limitations with hostPath volume mounts to the container runtime
+socket, it would have its own disadvantages that we are trying to avoid. For example, more permissions to schedule scan
+Jobs and additional information about cluster's infrastructure such as location of the container runtime socket. 
+
+## Solution
+
+### TL;DR;
+
+Use Trivy filesystem scanning to scan container images. The main idea, which is discussed in this proposal, is to
+schedule a scan Job on the same cluster node where the scanned workload. This allows Trivy to scan a filesystem of
+the container image which is already cached on that node without pulling the image from a remote registry. What's more,
+Trivy will scan container images from private registries without providing registry credentials (as ImagePullSecret or
+in any other proprietary way).
+
+### Deep Dive
+
+To scan a container image of a given K8s workload Starboard will create a corresponding container of a scan Job and
+override its entrypoint to invoke Trivy filesystem scanner.
+
+This approach requires Trivy executable to be downloaded and made available to the entrypoint. We'll do that by adding
```

Comment:
This could maybe be done with a csi driver as well.

Reply:
I'm not sure that I understood this comment. Could you elaborate on how we can use csi driver in this case?

Reply:
just thinking, instead of copying the file from an init container to an emtpydir, a csi driver in ephemeral mode could mount the file into the container.

Reply:
May not be worth the effort. but could be like:

kind: Pod
apiVersion: v1
metadata:
  name: my-csi-app-inline-volume
spec:
  containers:
    - name: my-frontend
      image: busybox
      command: [ "sleep", "100000" ]
      volumeMounts:
      - mountPath: "/trivy"
        name: my-csi-volume
  volumes:
  - name: my-csi-volume
    csi:
      driver: trivy